pkg/calculator/experimentcalc/experiment_calculator.go

```diff
@@ -0,0 +1,673 @@
+					break
+				}
+				time.Sleep(50 * time.Millisecond)
+			}
+			result, err := e.httpStan.GetFitResult(ctx, e.modelID, fitId)
+			if err != nil {
+				e.logger.Error("Failed to get fit result",
+					log.FieldsFromImcomingContext(ctx).AddFields(
+						zap.String("fitId", fitId),
+						zap.Error(err),
+					)...,
+				)
+				return
+			}
+			fit := e.httpStan.ExtractFromFitResult(ctx, result)
+			constrainedNames, _ := e.httpStan.StanParams(ctx, e.modelID, req.Data)
+			samplesChan <- fit.Select(constrainedNames)
+		}(i)
+	}
+	go func() {
+		wg.Wait()
+		close(samplesChan)
+	}()
+
+	samples := make([]dataframe.DataFrame, 0, numOfChains)
+	for sample := range samplesChan {
+		samples = append(samples, sample)
+	}
+
+	if len(samples) != numOfChains {
+		e.logger.Error("Failed to get all samples",
+			log.FieldsFromImcomingContext(ctx).AddFields(
+				zap.Int("numOfChains", numOfChains),
+				zap.Int("numOfSamples", len(samples)),
+			)...,
+		)
+		return nil, failedToSampleErr
+	}
+
+	variationResults := convertFitSamples(samples, vids, baseLineIdx)
+
+	calculationHistogram.WithLabelValues(binomialModelSampleMethod).Observe(time.Since(startTime).Seconds())
+	return variationResults, nil
+}
+
+func (e ExperimentCalculator) updateExperimentStatus(
+	ctx context.Context,
+	namespace string,
+	ex *experiment.Experiment,
+	now time.Time,
+) {
+	if ex.Status == experiment.Experiment_RUNNING ||
+		ex.Status == experiment.Experiment_WAITING {
+		twoDaysAgo := now.Add(-2 * 24 * time.Hour).Unix()
+		if ex.StopAt < twoDaysAgo {
+			req := &experiment.FinishExperimentRequest{
+				EnvironmentNamespace: namespace,
+				Id:                   ex.Id,
+				Command:              &experiment.FinishExperimentCommand{},
+			}
+			_, err := e.experimentClient.FinishExperiment(ctx, req)
+			if err != nil {
+				e.logger.Error("Failed to finish experiment",
+					log.FieldsFromImcomingContext(ctx).AddFields(
+						zap.String("experimentId", ex.Id),
+						zap.Error(err),
+					)...,
+				)
+			}
+			return
+		}
+	}
+	if ex.Status == experiment.Experiment_WAITING {
+		if ex.StartAt <= now.Unix() {
+			req := &experiment.StartExperimentRequest{
+				EnvironmentNamespace: namespace,
+				Id:                   ex.Id,
+				Command:              &experiment.StartExperimentCommand{},
+			}
+			_, err := e.experimentClient.StartExperiment(ctx, req)
+			if err != nil {
+				e.logger.Error("Failed to start experiment",
+					log.FieldsFromImcomingContext(ctx).AddFields(
+						zap.String("experimentId", ex.Id),
+						zap.Error(err),
+					)...,
+				)
+			}
+			return
+		}
+	}
+}
+
+//-------------------------------------utility functions----------------------------------------------
+
+func listEndAt(startAt, endAt, now int64) []int64 {
+	var timestamps []int64
+	if endAt >= now {
+		endAt = now
+	}
+	for ts := startAt + day; ts < endAt; ts += day {
+		timestamps = append(timestamps, ts)
+	}
+	timestamps = append(timestamps, endAt)
+	return timestamps
+}
```

Comment:
I don't know if this function works well with UI.(I wanna ask myself back in time)
If the server runs in UCT and a user uses a browser in GMT, for example, how does the UI display the counts?
The server side here looks good to me, though the UI should be checked.

Reply:
I think we should use GMT timezone for all experiments in MySQL, the UI can show the real time according to the user's browser.

Reply:
@Ubisoft-potato 
For new services, we are adding a new config called `timezone` in the helm chart so that the user can set his timezone.
Can you implement the timezone flag in the calculator service?

Reference
https://github.com/bucketeer-io/bucketeer/blob/main/pkg/eventcounter/cmd/server/server.go#L114

https://github.com/bucketeer-io/bucketeer/blob/main/pkg/eventcounter/api/api.go#L624

Reply:
>For new services, we are adding a new config called timezone in the helm chart so that the user can set his timezone.
>Can you implement the timezone flag in the calculator service?

@cre8ivejp No problem! Using timezone is a good solution for this.

Reply:
> @Ubisoft-potato For new services, we are adding a new config called `timezone` in the helm chart so that the user can set his timezone. Can you implement the timezone flag in the calculator service?

Done! And Helm Chart PR also added timezone env.

Reference:
https://github.com/bucketeer-io/bucketeer/pull/420